File: pkg/protocol/suite.go
package protocol

import "fmt"

// A Suite names a coherent set of algorithms for one device population.
//
// The wire format carries algorithm identifiers per message (SignedEnvelope.alg,
// SealedPayload.alg), so a server can serve several suites at once and a device
// only ever has to implement the one its fleet was provisioned with. The suite
// is what an operator selects — per profile, so a fleet of constrained devices
// can be served differently from a fleet of Linux boxes without a global flag
// day.
//
// Why a second suite exists: Mbed TLS — and therefore the PSA crypto stack that
// every Zephyr microcontroller build uses — does not implement Ed25519. A
// microcontroller that already carries P-256, SHA-256 and an AEAD for its own
// device-management TLS can reuse them at no extra flash cost, whereas Ed25519
// would mean a second signature stack and Curve25519 a second curve. On boards
// where the provisioning image already fills its flash partition, that
// difference decides whether the device can be provisioned at all.
type Suite string

const (
	// SuiteEd25519X25519 is the original suite and remains the default: every
	// existing agent (Go, Rust, POSIX shell) speaks it and nothing about those
	// code paths changes.
	SuiteEd25519X25519 Suite = "ed25519-x25519"

	// SuiteP256 is the constrained-device suite: one curve (NIST P-256) for
	// both signatures and key agreement, SHA-256 throughout, and the same
	// ChaCha20-Poly1305 AEAD as the default suite.
	SuiteP256 Suite = "p256"

	// DefaultSuite is used when neither the profile nor the server config
	// selects one.
	DefaultSuite = SuiteEd25519X25519
)

// Algorithm identifiers as they appear in SignedEnvelope.alg.
const (
	AlgEd25519   = "ed25519"
	AlgECDSAP256 = "ecdsa-p256-sha256"
)

// SignAlg returns the identifier this suite puts in SignedEnvelope.alg.
func (s Suite) SignAlg() string {
	if s == SuiteP256 {
		return AlgECDSAP256
	}
	return AlgEd25519
}

// SealAlg returns the identifier this suite puts in SealedPayload.alg and
// EncryptedPayload.alg.
func (s Suite) SealAlg() string {
	if s == SuiteP256 {
		return AlgP256ChaCha20
	}
	return AlgX25519ChaCha20
}

// EphemeralField names the EnrollRequest field a device of this suite uses to
// publish its ephemeral key-agreement public key. Used in error messages so an
// operator who has mismatched a profile against a device sees which field was
// expected.
func (s Suite) EphemeralField() string {
	if s == SuiteP256 {
		return "ephemeral_p256"
	}
	return "ephemeral_x25519"
}

// ParseSuite validates an operator-supplied suite name. The empty string
// selects DefaultSuite so that an omitted `crypto:` block means "unchanged".
func ParseSuite(s string) (Suite, error) {
	switch Suite(s) {
	case "":
		return DefaultSuite, nil
	case SuiteEd25519X25519:
		return SuiteEd25519X25519, nil
	case SuiteP256:
		return SuiteP256, nil
	default:
		return "", fmt.Errorf("unknown crypto suite %q (want %q or %q)",
			s, SuiteEd25519X25519, SuiteP256)
	}
}

// SuiteForSignAlg maps an inbound envelope's algorithm back to its suite. The
// server uses it to answer a device in the same suite it spoke, without the
// device having to name one.
func SuiteForSignAlg(alg string) (Suite, error) {
	switch alg {
	case AlgEd25519:
		return SuiteEd25519X25519, nil
	case AlgECDSAP256:
		return SuiteP256, nil
	default:
		return "", fmt.Errorf("unsupported alg %q", alg)
	}
}

File: clients/rust/src/suite.rs
//! Crypto suites — mirrors pkg/protocol/suite.go.
//!
//! A suite names a coherent set of algorithms. The server selects one per
//! profile; the agent picks the one its identity key was created for, signs its
//! EnrollRequest with it and publishes the matching ephemeral key-agreement key.
//! Inbound envelopes and sealed payloads always carry their own `alg`, so the
//! agent decides how to verify/decrypt from the message, never from config.

use std::fmt;
use std::str::FromStr;

/// `SignedEnvelope.alg` for Ed25519.
pub const ALG_ED25519: &str = "ed25519";
/// `SignedEnvelope.alg` for ECDSA P-256 / SHA-256 (raw `r‖s` signatures).
pub const ALG_ECDSA_P256: &str = "ecdsa-p256-sha256";

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum Suite {
    /// The original suite and the default: Ed25519 signatures, X25519 sealing.
    #[default]
    Ed25519X25519,
    /// NIST P-256 for both signatures and key agreement (HKDF-SHA256 derived
    /// ChaCha20-Poly1305 key). Selected per profile on the server.
    P256,
}

impl Suite {
    pub fn as_str(&self) -> &'static str {
        match self {
            Suite::Ed25519X25519 => "ed25519-x25519",
            Suite::P256 => "p256",
        }
    }

    /// Identifier this suite puts in `SignedEnvelope.alg`.
    pub fn sign_alg(&self) -> &'static str {
        match self {
            Suite::Ed25519X25519 => ALG_ED25519,
            Suite::P256 => ALG_ECDSA_P256,
        }
    }

    /// Identifier this suite puts in `SealedPayload.alg` / `EncryptedPayload.alg`.
    pub fn seal_alg(&self) -> &'static str {
        match self {
            Suite::Ed25519X25519 => crate::encrypt::ALG,
            Suite::P256 => crate::encrypt::ALG_P256,
        }
    }

    /// Maps an envelope's `alg` back to its suite.
    pub fn for_sign_alg(alg: &str) -> crate::Result<Suite> {
        match alg {
            ALG_ED25519 => Ok(Suite::Ed25519X25519),
            ALG_ECDSA_P256 => Ok(Suite::P256),
            other => Err(format!("unsupported alg {other:?}").into()),
        }
    }
}

impl fmt::Display for Suite {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

impl FromStr for Suite {
    type Err = String;

    /// The empty string selects the default, so an unset option means "unchanged".
    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
        match s {
            "" | "ed25519-x25519" => Ok(Suite::Ed25519X25519),
            "p256" => Ok(Suite::P256),
            other => Err(format!(
                "unknown crypto suite {other:?} (want \"ed25519-x25519\" or \"p256\")"
            )),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse() {
        assert_eq!("".parse::<Suite>().unwrap(), Suite::Ed25519X25519);
        assert_eq!("ed25519-x25519".parse::<Suite>().unwrap(), Suite::Ed25519X25519);
        assert_eq!("p256".parse::<Suite>().unwrap(), Suite::P256);
        assert!("rsa".parse::<Suite>().is_err());
    }

    #[test]
    fn alg_roundtrip() {
        for s in [Suite::Ed25519X25519, Suite::P256] {
            assert_eq!(Suite::for_sign_alg(s.sign_alg()).unwrap(), s);
        }
    }
}
